feat(MDXComponents): migrate markdown package to @carbon-labs

IgnacioBecerra opened the pull request:
Closes #509

#### Changelog

**New**

- {{new thing}}

**Changed**

- {{change thing}}

**Removed**

- {{removed thing}}

#### Testing / Reviewing

{{ Add descriptions, steps or a checklist for how reviewers can verify this PR works or not }}

mattrosno commented:
> Should we put this in its own top category so it doesn't get mixed in with the Labs components?

@jeffchew @IgnacioBecerra it's a package of components, that might not live here forever, but I'm game to isolate these since the intent isn't to use this in-product like the rest of the components.

mattrosno commented on packages/react/.storybook/preview.js:
```diff
@@ -190,14 +191,28 @@ const decorators = [
       setRandomKey(Math.floor(Math.random() * 10));
     }, [locale, dir]);
 
+    const markdownComponentsStyles = {
+      background: 'var(--cds-background)',
+      padding: '1rem',
+    };
```
@IgnacioBecerra I'll open a PR against this branch as an alternative way to do this.

mattrosno commented on packages/react/src/components/MDXComponents/components/_utils.scss:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright IBM Corp. 2022, 2023
```
```suggestion
 * Copyright IBM Corp. 2022, 2025
```

Can the last date in every copyright notice in the PR be 2025? First date is when the file was created, second date is the year in which the last edit was made.

IgnacioBecerra replied:
Yep, just pushed this in the last commit, was leaving it to the end lol 👍

mattrosno commented on packages/react/src/components/MDXComponents/components/art-direction/art-direction.stories.jsx:
```diff
@@ -0,0 +1,35 @@
+
+import React from 'react'
+import { ArtDirection } from './art-direction'
+import desktop from '../../../../../.storybook/storybook-images/desktop.jpg'
+import mobile from '../../../../../.storybook/storybook-images/mobile.jpg'
+import tablet from '../../../../../.storybook/storybook-images/tablet.jpg'
```
How can we do it so these images live next to its story? We'll need this for documentation page images too, so we don't have to put all images in a global images folder.

IgnacioBecerra commented on packages/react/src/components/MDXComponents/components/link/_link.scss:
```diff
@@ -0,0 +1,12 @@
+@use '../utils' as *;
+
+.#{with-prefix('link')} {
+  font-size: inherit !important;
```
Without the `!important`, the styles for `@carbon/react`'s Link are applied to the component, meaning that it's font size gets decreased, at least in Storybook.

Similar to what happened to the UIShell and the accidental WhatsNew style overrides.

mattrosno replied:
Let's look into this in the final review. I'm hoping that we can bring over the styles as-is without changing.

mattrosno replied:
@IgnacioBecerra are you sure we need to add `!important`s? I'm curious why it wasn't needed before. Is there a better solution besides using important?

mattrosno commented on packages/react/src/components/MDXComponents/components/_utils.scss:
```diff
@@ -0,0 +1,18 @@
+$carbon-platform-prefix: 'carbon-platform-mdx-components';
+
+@function with-prefix($input) {
+  @return $carbon-platform-prefix + '--' + $input;
```
Let's do a different prefix, since this isn't in the `carbon-platform` repo anymore. How about `$carbon-mdx`?

mattrosno commented:
We need to document:

* Supported MDX components
* Meta component with `title` is required
* Doesn't support frontmatter, HTML comments, or markdown syntax for images
* For images, instead, import the file and use an `<img />` element
* MDX components not supported but used in other Carbon sites
  * `A11yStatus`

mattrosno commented on packages/react/.storybook/preview.js:
```diff
@@ -1,13 +1,119 @@
 import { Layout } from '@carbon/react/es/components/Layout';
 import { TextDirection } from '@carbon/react/es/components/Text';
+import { DocsContainer, Meta, Unstyled } from '@storybook/blocks';
+import {
+  H1,
+  H2,
+  H3,
+  H4,
+  H5,
+  H6,
+  P,
+  OL,
+  UL,
+  LI,
+  Link,
+  Blockquote,
+  PageTable,
+  Caption,
+  Code,
+  AnchorLinks,
+  AnchorLink,
+  Column,
+  DoDont,
+  DoDontRow,
+  GifPlayer,
+  PageDescription,
+  Row,
+  StorybookDemo,
+  Grid,
+  Tabs,
+  InlineNotification,
+  Tab,
+  ArtDirection,
+} from '../src/components/MDXComponents';
```
It looks like we're missing these imports, to map in the components object below.

* Accordion
* ArticleCard
* Aside
* CardGroup
* Code
* Divider
* ImageWrapper
* MiniCard
* Preview
* ResourceCard
* Title
* Video
* AccordionItem
* Button
* Layer

Also, did it work to do the Lerna local package where we import from `@carbon-labs/mdx-components` and we do the wildcard dependency in the package.json?

IgnacioBecerra replied:
i tried it, it seems to works natively

mattrosno commented on packages/react/.storybook/preview.js:
```diff
@@ -1,13 +1,119 @@
+  Tab,
+  ArtDirection,
+} from '../src/components/MDXComponents';
+import '../src/components/MDXComponents/components/index.scss';
```
Same with this. Does the `@carbon-labs/mdx-components` package export the index.scss file? Should we do the Lerna workspace import here or keep it as a relative path?

IgnacioBecerra replied:
same as above, it works natively

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
````diff
@@ -0,0 +1,64 @@
+
+
+## Directory Structure  
+- All `.mdx` files **must** be placed under:  
+  ```
+  src/components/ComponentName/docs/
+  ```
````
I don't think this is true any more. For example, the example component uses a `__docs__` directory. We can remove this. Our MDX gets used if it's an MDX file, and it doesn't have an attached Component Story Format (CSF) like the autodocs MDX have.

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
````diff
@@ -0,0 +1,64 @@
+/>
+```
+
+## Available MDX Components  
````
Add the rest of the MDX components here too, like Preview.

mattrosno replied:
```suggestion
## Available MDX components  
```

mattrosno reviewed (changes requested):
Some visual differences compared to the original MDX package:

The anchor links look smaller compared to https://platform.carbondesignsystem.com/?path=/docs/components-anchorlinks--default

<img width="310" alt="image" src="https://github.com/user-attachments/assets/7967096a-d07f-4402-bc80-c40a8b2baecf" />

Should the table headers be left aligned? https://platform.carbondesignsystem.com/?path=/docs/components-markdown-pagetable--default
<img width="543" alt="image" src="https://github.com/user-attachments/assets/f7dd3851-43c4-4a47-8db5-c44c32472977" />

mattrosno commented:
<img width="1125" alt="image" src="https://github.com/user-attachments/assets/71402d73-bbf9-46dd-a1a1-b2fe9b9a78e3" />

The color of the inline code background looks off, and this class name looks suspicious. It looks Storybook default, not Carbon. I wonder if we need to provide styles to code elements.

mattrosno commented:
<img width="1061" alt="image" src="https://github.com/user-attachments/assets/edf76d67-2f1b-4aba-951f-08198cbf30ac" />
I wonder if the left alignment of the ordered list is a bug, or just wrong from the original MDX (e.g. maybe needs wrapped in a grid.)

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
```diff
@@ -0,0 +1,64 @@
+<Meta title="MDX Components/MDX Documentation Guidelines" />
+# MDX Documentation Guidelines  
```
```suggestion
<Meta title="MDX Components/Guidelines" />
# Guidelines  
```
We don't need to duplicate "MDX Components" in this heading.

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
```diff
@@ -0,0 +1,64 @@
+as the ones found in the [Carbon Design System](carbondesignsystem.com) page.
+
+
+## Directory Structure  
```
```suggestion
## Meta component
```
We use sentence casing for headings, per https://carbondesignsystem.com/guidelines/content/writing-style/#capitalization

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
````diff
@@ -0,0 +1,64 @@
+  ```
+  src/components/ComponentName/docs/
+  ```
+- The `Meta` component must be included at the top of the file with a `title` in the format:  
````
```suggestion
- The `Meta` Storybook MDX component must be included at the top of the file with a `title`. This allows you to determine where the file is included in the side navigation, instead of it being placed based on directory path by default.  
```

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
````diff
@@ -0,0 +1,64 @@
+  <Meta title="ComponentName/Example" />
+  ```
+
+## Unsupported Features  
````
```suggestion
## Unsupported features  
```

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
```diff
@@ -0,0 +1,64 @@
+- **HTML comments** (e.g., `<!-- comment -->`)
+- **Markdown image syntax** (e.g., `![alt text](./image.png)`)
+
+## Using Images  
```
```suggestion
## Using images  
```

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
```diff
@@ -0,0 +1,64 @@
+- `InlineNotification`  
+- `ArtDirection`  
+
+## Unsupported MDX Components  
```
```suggestion
## Unsupported MDX components  
```

mattrosno commented on packages/react/src/components/MDXComponents/overview.mdx:
```diff
@@ -0,0 +1,64 @@
+- `ArtDirection`  
+
+## Unsupported MDX Components  
+The following component is **not** supported in this documentation,
+but note that it is used in other Carbon sites:
```
```suggestion
The following component are **not** currently supported in this documentation, although they may be available in Carbon websites:
```

mattrosno reviewed (changes requested):
Some wording suggestions to add clarity, and capitalization fixes per our content guidelines. You'll probably need to format the file after accepting changes to pass linting.

mattrosno commented:
@IgnacioBecerra I'm seeing the code background being too tall in the latest change? Also, check out the inline code in the tables... it looks off still.

<img width="2880" alt="image" src="https://github.com/user-attachments/assets/728f7c35-85ec-4ffc-a011-2ce60ee4cb36" />

This looks normal in Carbon Next, so instead of making styling changes, I'd look to ensure all of the proper styling made its way over from the carbon-platform repo?

https://next.carbondesignsystem.com/libraries/carbon-react/main/assets/accordion/style

mattrosno reviewed (changes requested):
<img width="1126" alt="image" src="https://github.com/user-attachments/assets/60b20478-7096-4aa1-b88e-7181564e30f8" />

It looks like the small lost its small styling? Compared to this: https://platform.carbondesignsystem.com/?path=/docs/components-anchorlinks--small

IgnacioBecerra commented:
@mattrosno Addressed all those issues! What happened with the `Code` component is that Storybook was using the default `code` markdown component as you suspected. We were only using `pre: Code` in `preview.js`, however, when adding `code: Code`, it would break as-is. MDXJS used to have an `InlineCode` component, but it [got removed some time ago](https://mdxjs.com/migrating/v2/#update-mdx-content:~:text=The%20special%20component%20name%20inlineCode%20was%20removed%2C%20we%20recommend%20to%20use%20pre%20for%20the%20block%20version%20of%20code%2C%20and%20code%20for%20both%20the%20block%20and%20inline%20versions). 

Currently, the platform's `Code` component uses `CodeSnippet` in `multiline` configuration, so when it doesn't receive the right props (when using inline markdown), it breaks. The solution is to create a different component of ours, `InlineCode`, and redirect `code: InlineCode`. With this, we see all the right styles coming in properly, with no need for extra styling :)

mattrosno reviewed (approved):
Looks good here! @alisonjoseph can you review?

alisonjoseph reviewed (changes requested):
This looks hreat!! Just a few, hopefully small, things 🙏

**Accordion**
40px height, should be 48px

**AnchorLinks**
Should split to two column if more than 4 items
<img width="746" alt="Screenshot 2025-03-19 at 10 19 50 AM" src="https://github.com/user-attachments/assets/2f37d29d-b834-4644-87a6-d765cb490d4f" />

**Video**
Remove Vimeo - we aren’t allowed to use, or at least remove the example

  **Caption**
Missing fullWidth prop

**Blockquote** 
Dash should be in-line with text
<img width="805" alt="Screenshot 2025-03-19 at 10 21 07 AM" src="https://github.com/user-attachments/assets/f504a99a-475c-43eb-a211-208de48bf81d" />


**Grid**
We should document both ways of writing the Grid - reference that it supports the Carbon grid or the Gatsby theme grid https://gatsby.carbondesignsystem.com/components/Grid/ Also remove reference to “Platform” in the description.

IgnacioBecerra commented:
@alisonjoseph Your comments have been addressed in the latest commit
